### ISSUE TITLE
Rate limit number of messages per second for streaming gRPC

### ISSUE THREAD

**Issue by yxue** [envoyproxy/envoy#23942]
*Title*: *Rate limit number of messages per second for streaming gRPC*

*Description*:

Envoy has the rate limit for the downstream connections. For the streaming gRPC, could Envoy rate limit the message rate per stream?

**Comment by yxue** [comment]
Could Envoy read the message in a slow rate and take advantage of the gRPC flow control? Envoy doesn't need to respond to the client. The client's send could get blocked until the message is consumed by the Envoy.

**Comment by kyessenov** [comment]
Yeah, that is plausible, but that's more of a flow control decision, so it's a soft push back, rather than hard 429.

**Comment by yxue** [comment]
Yeah, I was thinking about the soft push back. For hard 429, could Envoy return the grpc status to the client and reset the upstream connection?

**Comment by kyessenov** [comment]
Sure, we have grpc_stats filter that can count requests/responses without having the protobufs. This sounds more like a bandwidth rate limit, except using messages instead of bytes.

**Comment by yxue** [comment]
Yes, I think the rate limiter could be the combination of grpc_stats and leaky bucket. If the message rate limit can be achieved, the bandwidth rate limit could be similar?

**Comment by qiwzhang** [comment]
Please leave this issue open.   I will do some prototype for this feature.  You can assign it to me
